packages/controllers/src/snaps/endowments/transaction-insight.ts

```diff
@@ -30,12 +45,117 @@ const specificationBuilder: PermissionSpecificationBuilder<
   return {
     permissionType: PermissionType.Endowment,
     targetKey: permissionName,
-    allowedCaveats: null,
+    allowedCaveats: [SnapCaveatType.SnapTransactionInsight],
     endowmentGetter: (_getterOptions?: EndowmentGetterParams) => undefined,
+    validator: ({ caveats }) => {
+      if (
+        caveats !== null &&
+        (caveats?.length > 1 ||
+          caveats[0].type !== SnapCaveatType.SnapTransactionInsight)
+      ) {
+        throw ethErrors.rpc.invalidParams({
+          message: `Expected a single "${SnapCaveatType.SnapTransactionInsight}" caveat.`,
+        });
+      }
+    },
   };
 };
 
 export const transactionInsightEndowmentBuilder = Object.freeze({
   targetKey: permissionName,
   specificationBuilder,
 } as const);
+
+/**
+ * Validates the type of the caveat value.
+ *
+ * @param caveat - The caveat to validate.
+ * @throws If the caveat value is invalid.
+ */
+function validateCaveat(caveat: Caveat<string, any>): void {
+  if (!hasProperty(caveat, 'value') || !isPlainObject(caveat.value)) {
+    throw ethErrors.rpc.invalidParams({
+      message: 'Expected a plain object.',
+    });
+  }
+
+  const { value } = caveat;
+
+  if (!hasProperty(value, 'allowTransactionOrigin')) {
+    throw ethErrors.rpc.invalidParams({
+      message: 'Expected a plain object.',
+    });
+  }
+
+  assert(
+    typeof value.allowTransactionOrigin === 'boolean',
+    'Expected allowTransactionOrigin to have type "boolean"',
+  );
+}
+
+/**
+ * Map a raw value from the `initialPermissions` to a caveat specification.
+ * Note that this function does not do any validation, that's handled by the
+ * PermissionsController when the permission is requested.
+ *
+ * @param value - The raw value from the `initialPermissions`.
+ * @returns The caveat specification.
+ */
+export function getTransactionInsightCaveatMapper(
+  value: Json,
+): Pick<PermissionConstraint, 'caveats'> {
+  if (!value || (isObject(value) && Object.keys(value).length === 0)) {
+    return { caveats: null };
+  }
+  return {
+    caveats: [
+      {
+        type: SnapCaveatType.SnapTransactionInsight,
+        value,
+      },
+    ],
+  };
+}
+
+export type TransactionInsightCaveat = {
+  allowTransactionOrigin?: boolean;
+};
```

Comment:
Rather than `allowTransactionOrigin`, simply the presence of a caveat (e.g., "TransactionOrigin") could be enough?

Reply:
Yeah potentially, do you think we should just use `allowTransactionOrigin`  in the manifest and then map it to having a caveat?

Reply:
Yeah. Currently we're just using a single caveat for everything, but there's nothing that says a permission couldn't have two or more caveats in the future, as long as we map it properly in the manifest.

Reply:
Done!